File: apps/flutter_parent/lib/screens/not_a_parent_screen.dart
// Copyright (C) 2019 - present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter_parent/l10n/app_localizations.dart';
import 'package:flutter_parent/network/utils/api_prefs.dart';
import 'package:flutter_parent/parent_app.dart';
import 'package:flutter_parent/router/panda_router.dart';
import 'package:flutter_parent/utils/common_widgets/empty_panda_widget.dart';
import 'package:flutter_parent/utils/design/parent_colors.dart';
import 'package:flutter_parent/utils/features_utils.dart';
import 'package:flutter_parent/utils/quick_nav.dart';
import 'package:flutter_parent/utils/service_locator.dart';
import 'package:flutter_parent/utils/url_launcher.dart';
import 'package:flutter_svg/svg.dart';

class NotAParentScreen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: <Widget>[
          Expanded(
            child: EmptyPandaWidget(
              svgPath: 'assets/svg/panda-book.svg',
              title: L10n(context).notAParentTitle,
              subtitle: L10n(context).notAParentSubtitle,
              buttonText: L10n(context).returnToLogin,
              onButtonTap: () async {
                await ApiPrefs.performLogout(app: ParentApp.of(context));
                await FeaturesUtils.performLogout();
                locator<QuickNav>().pushRouteAndClearStack(context, PandaRouter.login());
              },
            ),
          ),
          _appOptions(context),
        ],
      ),
    );
  }

  Padding _appOptions(BuildContext context) {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 32),
      child: ExpansionTile(
        trailing: SizedBox(width: 1), // Replace dropdown chevron with empty, non-zero width element
        title: Padding(
          padding: const EdgeInsetsDirectional.only(start: 16), // Offset to keep text centered
          child: Text(
            L10n(context).studentOrTeacherTitle,
            textAlign: TextAlign.center,
            style: Theme.of(context).textTheme.bodySmall,
          ),
        ),
        children: [
          Text(
            L10n(context).studentOrTeacherSubtitle,
            textAlign: TextAlign.center,
            style: Theme.of(context).textTheme.bodySmall,
          ),
          SizedBox(height: 24),
          _appButton(context, L10n(context).studentApp, L10n(context).canvasStudentApp, ParentColors.studentApp, 'assets/svg/canvas-logo-student.svg', () {
            locator<UrlLauncher>().launch('market://details?id=com.instructure.candroid');
          }),
          _appButton(context, L10n(context).teacherApp, L10n(context).canvasTeacherApp, ParentColors.teacherApp, 'assets/svg/canvas-logo-teacher.svg', () {
            locator<UrlLauncher>().launch('market://details?id=com.instructure.teacher');
          }),
          SizedBox(height: 24),
        ],
      ),
    );
  }

  Widget _appButton(BuildContext context, String name, String label, Color color, String logo, GestureTapCallback onTap) {
    return InkWell(
      onTap: onTap,
      child: Semantics(
        label: label,
        excludeSemantics: true,
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              SvgPicture.asset(
                logo,
                height: 48,
                color: color,
              ),
              SizedBox(width: 8),
              Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: <Widget>[
                  SvgPicture.asset(
                    'assets/svg/canvas-wordmark.svg',
                    alignment: Alignment.centerLeft,
                    height: 24,
                  ),
                  Text(
                    name,
                    style: TextStyle(color: color, fontSize: 10, fontWeight: FontWeight.w900),
                  ),
                ],
              )
            ],
          ),
        ),
      ),
    );
  }
}

File: apps/parent/src/main/java/com/instructure/parentapp/features/notaparent/NotAParentScreen.kt
/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.parentapp.features.notaparent

import androidx.annotation.DrawableRes
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.instructure.pandautils.compose.composables.EmptyContent
import com.instructure.parentapp.R


@Composable
internal fun NotAParentScreen(
    returnToLoginClick: () -> Unit,
    onStudentClick: () -> Unit,
    onTeacherClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Surface(
        color = colorResource(id = R.color.backgroundLightest),
        modifier = modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
    ) {
        var bottomExpanded by remember { mutableStateOf(false) }

        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            modifier = Modifier.padding(horizontal = 16.dp)
        ) {
            Spacer(modifier = Modifier.weight(1f))
            EmptyContent(
                emptyTitle = stringResource(id = R.string.notAParentTitle),
                emptyMessage = stringResource(id = R.string.notAParentSubtitle),
                imageRes = R.drawable.ic_panda_book,
                buttonText = stringResource(id = R.string.returnToLogin),
                buttonClick = returnToLoginClick
            )
            Spacer(modifier = Modifier.weight(1f))
            Spacer(modifier = Modifier.height(16.dp))
            if (bottomExpanded) {
                Spacer(modifier = Modifier.height(16.dp))
                Divider(
                    color = colorResource(id = R.color.backgroundMedium),
                    thickness = 1.dp
                )
                Spacer(modifier = Modifier.height(16.dp))
            }
            Text(
                text = stringResource(id = R.string.studentOrTeacherTitle),
                color = colorResource(id = R.color.textDarkest),
                fontSize = 12.sp,
                modifier = Modifier.clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null
                ) {
                    bottomExpanded = !bottomExpanded
                }
            )
            Spacer(modifier = Modifier.height(16.dp))
            AppOptions(bottomExpanded, onStudentClick, onTeacherClick)
        }
    }
}

@Composable
private fun AppOptions(
    bottomExpanded: Boolean,
    onStudentClick: () -> Unit,
    onTeacherClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = modifier
            .animateContentSize()
            .height(if (bottomExpanded) 180.dp else 0.dp)
    ) {
        Text(
            text = stringResource(id = R.string.studentOrTeacherSubtitle),
            color = colorResource(id = R.color.textDarkest),
            fontSize = 12.sp,
            textAlign = TextAlign.Center
        )
        Spacer(modifier = Modifier.height(16.dp))
        AppOption(
            stringResource(id = R.string.studentApp),
            stringResource(id = R.string.canvasStudentApp),
            colorResource(id = R.color.login_studentAppTheme),
            R.drawable.ic_canvas_logo_student,
            { onStudentClick() }
        )
        Spacer(modifier = Modifier.height(12.dp))
        AppOption(
            stringResource(id = R.string.teacherApp),
            stringResource(id = R.string.canvasTeacherApp),
            colorResource(id = R.color.login_teacherAppTheme),
            R.drawable.ic_canvas_logo_teacher,
            { onTeacherClick() }
        )
    }
}

@Composable
private fun AppOption(
    name: String,
    label: String,
    color: Color,
    @DrawableRes iconRes: Int,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier
            .clickable(onClick = onClick)
            .semantics(mergeDescendants = true) {
                contentDescription = label
            }
    ) {
        Icon(
            painter = painterResource(id = iconRes),
            tint = color,
            contentDescription = null,
            modifier = Modifier.size(48.dp)
        )
        Spacer(modifier = Modifier.width(8.dp))
        Column {
            Icon(
                painter = painterResource(id = R.drawable.ic_canvas_wordmark),
                tint = colorResource(id = R.color.backgroundMedium),
                contentDescription = null,
                modifier = Modifier.height(24.dp)
            )
            Text(
                text = name,
                color = color,
                fontSize = 10.sp,
                textAlign = TextAlign.Center
            )
        }
    }
}

@Preview
@Composable
fun NotAParentScreenPreview() {
    NotAParentScreen(
        returnToLoginClick = {},
        onStudentClick = {},
        onTeacherClick = {}
    )
}